### ISSUE TITLE
bug in hardware setup variables table

### ISSUE THREAD

**Issue by ThomasAkam** [pyControl/code#87]
While fixing a bug in 50bd9c11b4d62a2dbf22c14bb9248fb190494929 where running any task or experiment without naming the setups gave an error message, I encountered a bug in the hardware variables dialog.  The bug happens when I open the hardware variables dialog in the setups tab, and causes the error message below to appear in the log:

```
  File "E:\Dropbox\Hardware development\pyControl\gui\hardware_variables_dialog.py", line 113, in refresh_save_button
    if self.get_table_data() != self.starting_table:
  File "E:\Dropbox\Hardware development\pyControl\gui\hardware_variables_dialog.py", line 121, in get_table_data
    setup_name = self.cellWidget(table_row, 0).text()
AttributeError: 'NoneType' object has no attribute 'text'
```

I could not work out what is causing this, @alustig3 could you take a look.

Also,  I have a couple of questions/comments about the code for the `VariablesTable` widget in `hardware_variables_dialog.py`:

- At line [136](https://github.com/pyControl/code/blob/50bd9c11b4d62a2dbf22c14bb9248fb190494929/gui/hardware_variables_dialog.py#L136) what is the purpose of checking `if setup_variables is not False`, and could we refactor it to avoid this somewhat weird syntax?
- At line [130](https://github.com/pyControl/code/blob/50bd9c11b4d62a2dbf22c14bb9248fb190494929/gui/hardware_variables_dialog.py#L130), it is confusing to assign `setup_variables["hw_var"] = self.setup_var_editor.variable_cbox.currentText()` given that all the other entries in the `setup_variables` dictionary are `{setup_name: variable_value}`.  Mixing these different bits of information in the same dictionary then makes the logic of the code in the [save](https://github.com/pyControl/code/blob/50bd9c11b4d62a2dbf22c14bb9248fb190494929/gui/hardware_variables_dialog.py#L134-L151) method hard to follow.  Would it be possible to refactor this, e.g. by just reading the `variable_cbox.currentText()` when the save button is pressed?

**Comment by alustig3** [comment]
> where running any task or experiment without naming the setups gave an error message, I encountered a bug in the hardware variables dialog. The bug happens when I open the hardware variables dialog in the setups tab

I'm not sure I follow. would it be possible to share a screen recording of you encountering the error? You shouldn't be able to open the hardware variables dialog in the setups tab without first naming the setup (the `Variables` button should not be enabled)

![CleanShot 2023-06-02 at 15 21 31](https://github.com/pyControl/code/assets/8128628/8a74944f-b2f3-4d6d-8771-9ec97033ff08)

**Comment by ThomasAkam** [comment]
@alustig3 I get this error even when all the setups have names, see the configuration in the image below.  The error occurs when I open the hardware variables dialog, and only happens if there are saved values for hardware variables.  The `setups.json` file read:

```
{
    "COM4": {
        "name": "b2",
        "variables": {
            "hw_solenoid": 10
        }
    },
    "COM6": {
        "name": "b1",
        "variables": {
            "hw_solenoid": 8
        }
    }
}
```

![image](https://github.com/pyControl/code/assets/4661184/7115de88-6bf3-4d05-87af-4fdde37fbe94)
